providers/implementations/kdfs/snmpkdf.c

```diff
@@ -0,0 +1,314 @@
+#include "crypto/evp.h"
+#include "prov/provider_ctx.h"
+#include "prov/providercommon.h"
+#include "prov/implementations.h"
+#include "prov/provider_util.h"
+
+/* See RFC 3414, Appendix A.2.2 */
+/* See NIST SP800-135 Section 6.8 */
+static OSSL_FUNC_kdf_newctx_fn kdf_snmpkdf_new;
+static OSSL_FUNC_kdf_dupctx_fn kdf_snmpkdf_dup;
+static OSSL_FUNC_kdf_freectx_fn kdf_snmpkdf_free;
+static OSSL_FUNC_kdf_reset_fn kdf_snmpkdf_reset;
+static OSSL_FUNC_kdf_derive_fn kdf_snmpkdf_derive;
+static OSSL_FUNC_kdf_settable_ctx_params_fn kdf_snmpkdf_settable_ctx_params;
+static OSSL_FUNC_kdf_set_ctx_params_fn kdf_snmpkdf_set_ctx_params;
+static OSSL_FUNC_kdf_gettable_ctx_params_fn kdf_snmpkdf_gettable_ctx_params;
+static OSSL_FUNC_kdf_get_ctx_params_fn kdf_snmpkdf_get_ctx_params;
+
+#define KDF_SNMP_PASSWORD_HASH_AMOUNT (1024 * 1024)
+#define KDF_SNMP_MIN_PASSWORD_LEN 8
+
+static int SNMPKDF(const EVP_MD *evp_md,
+                   const unsigned char *eid, size_t eid_len,
+                   unsigned char *password, size_t password_len,
+                   unsigned char *okey, size_t okeylen);
+
+typedef struct {
+    void *provctx;
+    PROV_DIGEST digest;
+    unsigned char *eid;
+    size_t eid_len;
+    unsigned char *password;
+    size_t password_len;
+} KDF_SNMPKDF;
+
+static void *kdf_snmpkdf_new(void *provctx)
+{
+    KDF_SNMPKDF *ctx;
+
+    if (!ossl_prov_is_running())
+        return NULL;
+
+    if ((ctx = OPENSSL_zalloc(sizeof(*ctx))) != NULL)
+        ctx->provctx = provctx;
+    return ctx;
+}
+
+static void kdf_snmpkdf_free(void *vctx)
+{
+    KDF_SNMPKDF *ctx = (KDF_SNMPKDF *)vctx;
+
+    if (ctx != NULL) {
+        kdf_snmpkdf_reset(ctx);
+        OPENSSL_free(ctx);
+    }
+}
+
+static void kdf_snmpkdf_reset(void *vctx)
+{
+    KDF_SNMPKDF *ctx = (KDF_SNMPKDF *)vctx;
+    void *provctx = ctx->provctx;
+
+    ossl_prov_digest_reset(&ctx->digest);
+    OPENSSL_clear_free(ctx->eid, ctx->eid_len);
+    OPENSSL_clear_free(ctx->password, ctx->password_len);
+    memset(ctx, 0, sizeof(*ctx));
+    ctx->provctx = provctx;
+}
+
+static void *kdf_snmpkdf_dup(void *vctx)
+{
+    const KDF_SNMPKDF *src = (const KDF_SNMPKDF *)vctx;
+    KDF_SNMPKDF *dest;
+
+    dest = kdf_snmpkdf_new(src->provctx);
+    if (dest != NULL) {
+        if (!ossl_prov_memdup(src->eid, src->eid_len,
+                              &dest->eid, &dest->eid_len)
+                || !ossl_prov_memdup(src->password, src->password_len,
+                                     &dest->password , &dest->password_len)
+                || !ossl_prov_digest_copy(&dest->digest, &src->digest))
+            goto err;
+    }
+    return dest;
+
+ err:
+    kdf_snmpkdf_free(dest);
+    return NULL;
+}
+
+static int snmpkdf_set_membuf(unsigned char **dst, size_t *dst_len,
+                             const OSSL_PARAM *p)
+{
+    OPENSSL_clear_free(*dst, *dst_len);
+    *dst = NULL;
+    *dst_len = 0;
+    return OSSL_PARAM_get_octet_string(p, (void **)dst, 0, dst_len);
```

Comment:
I know this was copied from somewhere else, but can you see what happens for the cases of p->data_size = 0
OR p->data = NULL.

Reply:
I assume this was allowing some settings to optionally be set back to NULL. Probably not needed if the values need to be set.

Reply:
p->data = NULL triggers snmpkdf_set_membuf to return 0 indicating an error with the params
p->data_size = 0 is allowed